src/daemon/rpmostreed-transaction-types.c

```diff
@@ -458,6 +458,10 @@ rollback_transaction_execute (RpmostreedTransaction *transaction,
     return glnx_throw (error, "No rollback deployment found");
   else if (!rollback_deployment)
     {
+      if (ostree_deployment_is_staged (pending_deployment))
+        /* XXX: or just clean it up for them? also, awkward CLI mention */
+        return glnx_throw (error, "Staged deployment (remove with cleanup -p)");
```

Comment:
I always thought the "rollback to undo" was weird...feels like should be able to make a hard cutover, there'll be test fallout in a-h-t etc probably but that's just our tests.  Are we breaking anyone's *real world* workflow?  Probably not.

How about `No rollback target available (remove pending with cleanup -p)`?

Reply:
Now that `cleanup` exists, less likely yeah (though some folks might still be used to that workflow from pre-`cleanup` era). I think there's some value too in being able to keep a broken update as a second menu entry for testing/debugging? Even now that we have `ostree admin pin`, it seems like if we're in this situation

```
deployment 0: broken update
deployment 1: booted deployment
```

you'd have to do e.g. `ostree admin pin 0`, then `rpm-ostree deploy $booted_checksum`, right? Redeploying the booted deployment feels odd.

Reply:
Also today actually...shouldn't `ostree admin pin 0` fail on staged deployments?  (It may not...but that'd be a bug)

Reply:
> Also today actually...shouldn't `ostree admin pin 0` fail on staged deployments? (It may not...but that'd be a bug)

Good catch -- just tried this out, it doesn't!

Though I'm more talking about the broken update case. The context for my previous comment is that *before* `cleanup -p` came, doing `rollback` when booted in the second entry was somewhat part of the workflow: you boot into an update, find out it's broken, reboot into the previous tree, and then you want to make your booted tree the default again.

What I mean is that even now with `cleanup -p`, it's still useful to be able to just swap the order of existing deployments, rather than redeploying things just to change the order. A new `rpm-ostree edit` command would definitely drop the need for this -- I think there's an RFE for that... ahh: #910.

Reply:
OK I'll take the ostree-staged-pin bug.  And yep doing #910 would be cool for sure.